cmd/node.go

```diff
@@ -284,14 +371,139 @@ func (node *Node) Start(exitChannels ...*chan struct{}) {
+	// Set to STOPPED on calling *Node.Stop()
+	case STOPPED:
+		return STOPPED, nil
+	// Any other status code apart from the cases above are considered INVALID!!!!
+	default:
+		return INVALIDNODESTATUS, ErrInvalidNodeStatus
+	}
+}
+
+// Verifies whether a status code of the node is a valid status code.
+// Used while changing the status of the node
+// Modifying LoadStatus() and the validateStatus() functions and their tests are hard requirements
+// to add new status codes for the node.
+func validateNodeStatus(status NodeStatus) error {
+	switch status {
+	// Instance of the *Node is created, but not yet intialized using *Node.Start()
+	case NEVERSTARTED:
+		return nil
+	// Node is started using *Node.Start() and not stopped yet.
+	case RUNNING:
+		return nil
+	// Node was once initialized, but currently stopped.
+	// Set to STOPPED on calling *Node.Stop()
+	case STOPPED:
+		return nil
+	// Any other status code apart from the cases above are considered INVALID,
+	default:
+		return ErrInvalidNodeStatus
+	}
+}
+
+// changes the running status of the node
+// Always use this function to change the status of the node.
+func (node *Node) changeNodeStatus(newStatus NodeStatus) error {
+	if err := validateNodeStatus(newStatus); err != nil {
+		return err
+	}
+
+	// Cannot deliberately change the status to NEVERSTARTED.
+	// NEVERSTARTED is the default status of the node before the
+	// node is started using *Node.Start().
+	if newStatus == NEVERSTARTED {
+		return ErrCannotSetToNeverStarted
+	}
+	// Load the current status of the Node.
+	status, err := node.LoadStatus()
+	if err != nil {
+		return err
+	}
+
+	// Cannot change the status of the server to STOPPED while it was never initalized
+	// in the first place. Can stop the never only after it's started using *Node.Start().
+	// Valid status transition is NEVERSTARTED -> RUNNING -> STOPPED -> RUNNING
+	if status == NEVERSTARTED && newStatus == STOPPED {
+		return ErrNodeNeverStarted
+	}
+
+	// No need to change the status if the new status is same as current status.
+	if newStatus == status {
+		switch newStatus {
+		case RUNNING:
+			return ErrAlreadyStarted
+		case STOPPED:
+			return ErrAlreadyStopped
+		}
+	}
+
+	node.status = &newStatus
+
+	return nil
+}
+
+// Helper function to check if the node is running.
+// returns false if node status is not set to RUNNING
+func (node *Node) IsRunning() bool {
+	status, err := node.LoadStatus()
+
+	if err != nil {
+		return false
+	}
+	if status == RUNNING {
+		return true
+	}
+	return false
+}
+
 func (node *Node) Stop() {
 	node.runningMutex.Lock()
 	defer node.runningMutex.Unlock()
 
-	if !node.IsRunning {
+	// Change nodes running status to stop
+	// Node has to be in RUNNING status to be able to STOP!
+	// FIXME: Return error in cases where the the node fails to stop. This makes the function testable.
+	if err := node.UpdateStatusStopped(); err != nil {
+		glog.Errorf("Error stopping Node -- %v", err)
```

Comment:
seems like we glog this same statement twice. we should either do it in the `UpdateStatusStopped` or here. My preference is that the caller handles how to glog so we don't blow up our logs, although I don't believe we'd hit this error statement in production often